### ISSUE TITLE
Summary about project's requirements

### ISSUE THREAD

**Issue by tsirif** [mila-udem/orion#2]
Summary of 04/10/2017 discussion and other notes
===

We would like our design to be aimed at satisfying the following 3 general requirements with descending priority order:

1. being **framework agnostic**:
    The training procedure is completely specified in its own executable, which will be executed by a child process (?).
2. being **language agnostic**:
    Same reason as (1). We will begin implementing an API for *Python* due to popularity. A client/worker side API will be designed to enable communication with the database and the server/master.
3. being **database agnostic**:
    This goal is to be able to run this software in a variety of web services/platforms. Maybe we can find a *Python* OSS, which wraps many popular databases, to act as a facade. We agreed that the initial implementation will be made in *MongoDB*.

Clarification on what is meant by *complete specification* of a training procedure
---
 * It uses some ML framework (we shouldn't care actually if it does).
 * It fetches the training dataset and the evaluation dataset.
 * Defines a model with its *parameter space* and part of architectural *hparam space*.
 * Defines a training optimization procedure with the rest of *hparam space*.
 * It will use our API to declare which *hparams* it should look for and which training *stats* it would export to the database.

So, the general goal, if I am thinking this right, is to decouple the training procedure itself from the hyperoptimization procedure.

**Worker-side** responsibilities
---
 * The *hparams* should be imported through our interface, using a YAML, json or another type of file which is to be set by the hyperoptimization procedure, and they will be also recorded inside the database as metadata of a particular training run.
 * Training *stats* are to be pushed by calls to our API within a worker's code, which can also trigger hooks to some master's procedure.
 * Part of our software library, within the worker process, will internally perform some of the calculations required by the hyperoptimization process, because:
   1. some data structures will be available in the workers only, e.g. evaluation sets for the hyperoptimization objective.
   2. concurrency
* Internal to our library, appropriate or desired modules will be loaded for the hyperoptimization procedure:
   1. A *hyperoptimization objective function*, which could be performance (accuracy, etc) on an hyperopt evaluation dataset or some form of regret function or anything which possibly takes as input the *hparam space* and returns a scalar.
   2. A *hyperopt method*: random search, grid, gradient-based, bayesian, genetic etc.
* Support inter-worker asynchronous and synchronous communication (?): MPI (?), nccl (?), sockets, collective operations (probably useful: AllGather, Broadcast) 

**Master-side** (maybe daemon?) responsibilities
---
 * Setting which hyperoptimization method and function to be loaded in the workers, also account for default procedures or automated selections according to the hyperparameters declared to be exported by worker's code.
 * Broadcasting other options and stuff to workers :P 
 * Distributed locking mechanism orchestration, if needed?
 * Resetting communicator worlds or managing what happens when a worker dies or effectively logs in.

Please complete any other thing missing from the aforementioned and comment on any of those. We have discussed whether it is possible to reuse some of the other hyperoptimization frameworks available, e.g. [hyperopt](https://github.com/hyperopt/hyperopt).

Other hyperoptimization software in general
---
 * [hyperopt](https://github.com/hyperopt/hyperopt)
 * [hype](https://github.com/hypelib/Hype)
 * [spearmint (old)](https://github.com/JasperSnoek/spearmint)
 * [spearmint](https://github.com/HIPS/Spearmint)

**Comment by bouthilx** [comment]
Thanks for the summary @tsirif !

I do not believe the meta-opt should have to process any dataset of some kind. It should be enough to only have hyper-parameters with metrics provided by the user's script. Also, I think the worker should fetch the hyper-parameters from our DB interface, but would communicate those to the user's script through yaml files or whatever would make it easy and avoid adding dependency on any programming language.

**Comment by nouiz** [comment]
I don't think the user script can query our interface for what to report. I think it is the user responsability to implement that and that this should be in the documentation.

Yes we could provide it in the API and the user could automatically collect that, but this would request much more change to user code. So anything more then having in the doc what to report and requesting to user to modify his script to collect that and return that is for later.

**Comment by nouiz** [comment]
Also, there is no communication between workers. Only between workder and db for now. Later maybe between worker and server instead of db.

**Comment by bouthilx** [comment]
Adding to @nouiz's comment, in a case where the user is not providing to expected information (some kind of metrics, with specific names), the algorithm module should very clearly point out in its error message what is missing so that the user could fix it easily without searching around in the documentation.
